### ISSUE TITLE
[clang-tidy] Inconsistent behavior for cppcoreguidelines-pro-type-member-init check

### ISSUE THREAD

**Issue by MaxWipfli** [llvm/llvm-project#59847]
`clang-tidy` has inconsistent behavior on the `cppcoreguidelines-pro-type-member-init`. See the following example.

---

In all cases, the `class Foo` contains a `char m_first` member variable, which is left uninitialized by the default constructor. (I have added an accessor for `m_first` such that it would be reachable from arbitrary code outside this file, since I wasn't sure if this affects the results.)

**Case 1**

```cpp
class Foo {
public:
    char first() const { return m_first; }
private:
    char m_first;
};
```

```console
$ clang-tidy -checks='cppcoreguidelines-*' foo.hpp
Error while trying to load a compilation database:
[...]
Running without flags.
[end of clang-tidy execution]
```

Thus, the uninitialized member is not detected by `clang-tidy`.

**Case 2 (with uninitialized second member)**
```cpp
class Foo {
public:
    char first() const { return m_first; }
private:
    char m_first;
    char m_second;
};
```
```console
$ clang-tidy -checks='cppcoreguidelines-*' foo.hpp
Error while trying to load a compilation database:
[...]
Running without flags.
[end of clang-tidy execution]
```
Nothing changes, still not detected.

**Case 3 (with default-initialized second member)**
```cpp
class Foo {
public:
    char first() const { return m_first; }
private:
    char m_first;
    char m_second { 0 };
};
```
```console
$ clang-tidy -checks='cppcoreguidelines-*' foo.hpp
Error while trying to load a compilation database:
[...]
Running without flags.
1 warning generated.
/home/max/Downloads/foo.hpp:1:7: warning: constructor does not initialize these fields: m_first [cppcoreguidelines-pro-type-member-init]
class Foo {
      ^
```
Now, the error is detected.

**Case 4 (with explicitly declared `= default` constructor)**
```cpp
class Foo {
public:
    Foo() = default;
    char first() const { return m_first; }
private:
    char m_first;
    char m_second;
};
```
```console
$ clang-tidy -checks='cppcoreguidelines-*' foo.hpp
Error while trying to load a compilation database:
[...]
Running without flags.
[end of clang-tidy execution]
```
Not detected.

**Case 5 (with explicitly declared `Foo() {}` constructor)**
```cpp
class Foo {
public:
    Foo() {}
    char first() const { return m_first; }
private:
    char m_first;
    char m_second;
};
```
```console
$ clang-tidy -checks='cppcoreguidelines-*' foo.hpp
Error while trying to load a compilation database:
[...]
Running without flags.
1 warning generated.
/home/max/Downloads/foo.hpp:3:5: warning: constructor does not initialize these fields: m_first [cppcoreguidelines-pro-type-member-init]
    Foo() {}
    ^
```
Now it is detected (presumably because this is not "the default constructor", and thus handled differently.

**Summary**
The uninitialized member `m_first` is only detected if either
* there is another, default-initialized member in the same class, or
* there is a explicitly defined `Foo () {}` default constructor (`= default` is not enough).

----

I am using `clang-tidy` version 15.0.6, downloaded from the release page of this GitHub repository.
```console
$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 15.0.6
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: tigerlake
```
